### PR TITLE
feat!: improve support for async/await

### LINKED ISSUE #1791: Default command invoked when calling .showHelp
```javascript
const yargs = require('yargs')

yargs.command('$0', 'a default command', (yargs) => yargs, () => {
  console.log('default command!! why?')
})
yargs.showHelp((output) => {
  console.log(output)
})
```

The above application will also execute the default command even though we never called `.argv` or `.parse`.  This seems like unexpected behavior to me and I could not find any docs on this.  Is this intended for some reason I am unaware of?

### PR DESCRIPTION
**Pull request by bcoe**
In `node@15` [top level await](https://v8.dev/features/top-level-await) is now supported 🎉. 

This means that we can have yargs return a promise in more situations, with users writing code that looks like this:

```js
import yargs from './index.mjs';
import { setTimeout } from 'timers/promises';

const argv = await yargs(process.argv.slice(2))
  .command(
    'foo',
    'foo commnd',
    () => {},
    async (argv) => {
      console.info('second')
      await setTimeout(1000)
    }
  )
  .middleware(async (argv) => {
    console.info('first')
    await setTimeout(1000)
    argv.value = 99
  }).argv;
console.info('third', argv)
```

This PR begins work to add better first class support of promises to yargs.

---

TODO:

- [x] ~write unit tests for scenarios outlined in #1420~ broke into several steps.
- [x] add tests for async behavior of `.parse()` (Refs: #1420). 
- [x] add tests for async behavior of top level `.argv` (Refs: #1420).
- [x] add tests for, and document `async getHelp() (Refs: #1420).
- [x] add tests for async getCompletion()/completion (Refs: #1420, #1235).
- [x] write unit tests for any TODO comments added during this work.
- [x] ~should command builders accept a promise?~ see #1042 -- we will land this work separately before the next major.
- [x] ~add async support to coerce/check~, see #859 -- we will land this work separately before the next major.
- [x] ~make global middleware work without defining command (_this work should be in separate breaking PR_).~ broken into ticket https://github.com/yargs/yargs/issues/1849
- [x] add documentation for new async behavior.

BREAKING CHANGES:

* now returns a promise if handler is `async`.
* onFinishCommand removed, in favor of being able to await promise.
* getCompletion now invokes callback with `err` and `completions, returns promise of completions.

Fixes #1791, #1420, #1235

### CODE REVIEW

**Comment by bcoe** [comment]
For anyone who would like to try this and give feedback (_which is greatly appreciated_):

`npm i yargs@next`

**Comment by bcoe** [comment]
@mleguen @SkeLLLa @petrgrishin @dapplion this PR removes `onFinishCommand` in favor of a variety of work to make commands work better with promises. You can now simply `await` yargs when it has an async command.

I would love your feedback. You can install this version of yargs like so:

`npm i yargs@next`

**Comment by dapplion** [comment]
> @mleguen @SkeLLLa @petrgrishin @dapplion this PR removes `onFinishCommand` in favor of a variety of work to make commands work better with promises. You can now simply `await` yargs when it has an async command.
> 
> I would love your feedback. You can install this version of yargs like so:
> 
> `npm i yargs@next`

The overall direction looks great! Can't wait to have this on master

**Review by petrgrishin — APPROVED** [review]

File: docs/advanced.md [hunk 1]
```diff
@@ -567,3 +554,66 @@ var argv = require('yargs/yargs')(process.argv.slice(2))
      )
   .argv;
 ```
+
+## Using Yargs with Async/await
+
+If you use async middleware or async handlers for commands, `yargs.parse` and
+`yargs.argv` will return a Promise. When you `await` this promise the final
+parsed result will be returned:
```
Comment (bcoe) [inline docs/advanced.md:562]: Add a note that this will be returned after the handler has finished.

**Review by mdy970 — COMMENTED** [review]
Hi

**Comment by dossy** [comment]
I was surprised and disappointed by the removal of `onFinishCommand()`.

[`package.json`](https://github.com/yargs/yargs/blob/99016f6d63152e3f60b41e0e3ba3bbd55389e1c4/package.json#L113) indicates `node >= 12` which should have been changed to `node >= 15` if you're going to remove `onFinishCommand()` in yargs@17.

**Comment by jesec** [comment]
@bcoe The typing is wrong. See my comment in https://github.com/yargs/yargs/issues/2175#issuecomment-1120314276.

**Pull request merged**